### ISSUE TITLE
Why filename look like 256_65ad275398e897c8

### ISSUE THREAD

**Issue by BigBear0099** [thumbsviewer/thumbsviewer#3]
Hi！
I want to get the right file name for the picture, but why does it all look like 256_65ad275398e897c8

**Comment by erickutcher** [comment]
Those are what Windows (Vista and above) will store in Thumbs.db files in place of a filename. The only way to get the original file name is to scan the system with the Tools > Map Files Paths... menu. If the original file doesn't exist, then the name is lost forever.

**Comment by BigBear0099** [comment]
Wow! I understand now，i can get right file name in GUI right now，so how can i output right filename image in commanline?I mean, how do I use the ability to map file paths under the command line?

**Comment by erickutcher** [comment]
Unfortunately I don't have the mapping feature implemented for the command-line.

You could extract all the images in the GUI after they've been mapped by going to File > Save All...

You can also export the entries information as a CSV file by going to File > Export to CSV...

A CSV file can be loaded in any spreadsheet program like Microsoft Excel, or Google Sheets.

**Comment by BigBear0099** [comment]
Can you tell me how to convert the file ID to a file name? In Turn. can also get the file ID from file name? I want to read the information in the thumb.db in my Python program. so when I use the command line to extract images from the thumb.db. I can compare the file name to get the correct thumbnail. please give me a hint.

**Comment by erickutcher** [comment]
The names in the Thumbs.db files are hash values. You won't be able to convert them back into the original file name. The only way to know the original name is to either brute force them (which isn't feasible), or to find the original hash inputs by scanning the system. Luckily the input is just the filename.

I can't help with Python code, but here's some C++ code I wrote to do this:

```
unsigned long long hash_data( char *data, unsigned long long hash, short length )
{
	while ( length-- > 0 )
	{
		hash ^= ( ( ( hash * 0x820 ) + ( *data++ & 0x00000000000000FF ) ) + ( hash >> 2 ) );
	}

	return hash;
}

unsigned long long hash_file( wchar_t *filename )
{
	// Initial hash value. This value was found in thumbcache.dll.
	unsigned long long hash = 0x295BA83CF71232D9;

	// Hash the filename.
	return hash_data( ( char * )filename, hash, ( short )( wcslen( filename ) * sizeof( wchar_t ) ) );
}

```
The filename needs to be a wide character string. Don't include the NULL terminator when hashing.

Another thing to note is that the hash value in the Thumbs.db starts after the underscore character '\_'. For example, you'll want to ignore the "256_" in "256_65ad275398e897c8". "65ad275398e897c8" is the hash value.

**Comment by eocron** [comment]
Sorry for necroing, but found you on github. I need ability to create this hash name out of filename and using your sample, but in C#:

        //fileName=    bbb.jpg
        //encoding=   UnicodeEncoding.Unicode
        private static string GetHash(string fileName, Encoding encoding)
        {
            var ccc = encoding.GetBytes(fileName);
            ulong hash = 0x295BA83CF71232D9;
            for(int i = 0; i < ccc.Length; i++)
            {
                var c = (ulong)ccc[i];
                var tmp = (c & 0x00000000000000FF);
                hash ^= (hash * 0x820) + tmp + (hash >> 2);
            }

            return hash.ToString("X");
        }

Expected hash:  1404606d28f66d74
My function hash: AEA16768F30C5A62

Any thoughts as to why this hash not match? My OS is Windows 10 and I can provide my thumbcache.dll too if you want.
Also, there is some obscure bytes in header at offset 12 - as I understand it is checksum of sort? How I can replicate it?

Purpose of all this: I want ability to create Thumbs.db myself.

**Comment by RupW** [comment]
Where did you get the expected hash from? The algorithm posted still works for me with old thumbs.db files I've got, but nowadays Windows 10 stores all thumbnails in a different format in %LOCALAPPDATA%\Microsoft\Windows\Explorer. Is it those you're working with?

**Comment by thumbsviewer** [comment]
Wow! Thank you for this. I would not have known.

Windows 10 changed the initial hash value to 0x68DFB54498C54783

So change ulong hash = 0x295BA83CF71232D9; to ulong hash = 0x68DFB54498C54783; and it should give you the "1404606d28f66d74" expected hash.

I'll update the program to reflect this new information.

**Comment by eocron** [comment]
Changing hash worked, now I get correct hash =)

Any thoughts on what is inside checksum bytes at the end of the header? For example, for my file, named "bbb.jpg", here is its header info:

1800 0000 0300 0000 412d 0000 0000 0000 a967 8b23 fc09 2065 .... the rest is image data in JPEG format.

Also found this link which tells that can be possibly inside hashing algorithm for checksum - http://www.swiftforensics.com/2012/06/windows-7-thumbcache-hash-algorithm.html

Those last 8 bytes probably checksum of sort, because it changes if I change modification date of file or it's name.

**Comment by eocron** [comment]
> Where did you get the expected hash from? The algorithm posted still works for me with old thumbs.db files I've got, but nowadays Windows 10 stores all thumbnails in a different format in %LOCALAPPDATA%\Microsoft\Windows\Explorer. Is it those you're working with?

No, Im working with Thumbs.db which created on Network Shares (option is disable in Windows 10 by default, but can be reenabled). Those are still usefull in situtations where you have media which can be previewed, but absolutely can't and shouldn't be loaded by Explorer JUST to create thumbnails in its own cache (and probably hang for couple of minutes upon doing so). So Im writing an app which can provide Thumbs.db file on the flight for specified files (fortunatelly it is not computation high and only needed for batch of selected files).

**Comment by thumbsviewer** [comment]
Those 8 bytes are related to the file's date and time. I did find one instance of a Thumbs.db file where it contained a FILETIME value. Newer versions use something different and I'm not sure what it is. I'll have to look into it more.

Some other notes:
Windows Vista and 7 use 0x295BA83CF71232D9 as the initial hash value.
Windows 8, 8.1, and 10 use 0x68DFB54498C54783
I don't know about Windows 11, but it probably uses the same as 10.

**Comment by eocron** [comment]
Thank you for your work! I made great progress thanks to your repo on my own project!

**Comment by thumbsviewer** [comment]
Glad I could help. I'm still uncertain as to those 8 bytes. I used IDA to disassemble the thumbcache.dll and it appears that from Windows 8 and onward those 8 bytes are a hash value. Windows Vista and 7 don't calculate a hash and I'm pretty sure they just use the image's last modified FILETIME value. The problem I've had trying to figure this out is I can't debug it properly. If I could step though the assembly code, then it would be pretty easy, but the IDA debugger doesn't stop on any of the breakpoints I have set. Reverse engineering isn't really my specialty either.

**Comment by thumbsviewer** [comment]
I solve it.

Here's a C++ algorithm to calculate the value. It's based on the same code in Thumbs Viewer.

```
unsigned long long hash_date( wchar_t *filepath )
{
	// Initial hash value. This value was found in thumbcache.dll.
	unsigned long long hash = 0xBBA03D3F2F654661;

	HANDLE hFile = CreateFile( filepath, 0, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL );
	if ( hFile != INVALID_HANDLE_VALUE )
	{
		// Hash File ID - found in the Master File Table.
		BY_HANDLE_FILE_INFORMATION bhfi;
		GetFileInformationByHandle( hFile, &bhfi );
		CloseHandle( hFile );
		unsigned long long file_id = bhfi.nFileIndexHigh;
		file_id = ( file_id << 32 ) | bhfi.nFileIndexLow;

		hash = hash_data( ( char * )&file_id, hash, sizeof( unsigned long long ) );

		// Hash Last Modified FILETIME.
		return hash_data( ( char * )&bhfi.ftLastWriteTime, hash, sizeof( unsigned long long ) );
	}
	else
	{
		return 0;
	}
}
```

Thumbs.db are created on the client computer and then copied to the shared folder. So the filepath variable above is the actual location of the original file (the file whose thumbnail is created in Thumbs.db).
